bluetooth: add config/code to priorize ISO connections in tx processor

mjchen0 opened the pull request:
ISO data is time sensitive, and previous QoS logic had no prioritization between ACL and ISO data. This
change creates a dedicated ready list for ISO connections when ISO is enabled, and the tx processor always drains the ISO ready list first before processing non-ISO channels.

mjchen0 committed:
bluetooth: change to priorize ISO connections in tx processor

ISO data can be time sensitive, and previous QoS logic had
no prioritization between ACL and ISO data. This
change creates a dedicated ready list for ISO connections
when ISO is enabled, and the tx processor always
drains the ISO ready list first before processing
non-ISO channels.

ALso change the ISO bt_conn_tx free list so that ISO ones
are kept separate from the free list for non ISO connections.
They already came from separate pools, each pool sized by a
respective TX_BUF_COUNT, but they had been put into a common
free list. Having separate free lists prevents the starvation
of non-ISO connections by heavy ISO traffic. Once all
ISO bt_conn_tx buffers are used, matching the ISO_TX_BUF_COUNT,
the transmit from ready ISO connections will naturally be
paused until they are freed up and non-ISO connections will
get a chance to be processed.

Signed-off-by: Mike J. Chen <mjchen@google.com>

Thalley reviewed (commented):
I understand what you are trying to achieve here, and I'm not opposed to it as such. 

However your argument that ISO is more timing sensitive also applies to e.g. audio over BR/EDR, right? 

I'm wondering if, if we continue with different priorties per connection type, that we should do it in a more generic way that also accommodates the classic connection types, and any future possible types?

mjchen0 commented:
> I understand what you are trying to achieve here, and I'm not opposed to it as such.
> 
> However your argument that ISO is more timing sensitive also applies to e.g. audio over BR/EDR, right?
> 
> I'm wondering if, if we continue with different priorties per connection type, that we should do it in a more generic way that also accommodates the classic connection types, and any future possible types?

Unfortunately, I only have been using Zephyr with Nordic bluetooth/ble chips and don't have any BR/EDR chip to test with and am generally not that familiar with BR/EDR, so I'm not sure what a more generic change would look like. Any suggestions?

For another PR, I added some tests under tests/bsim, but those appear to only cover BLE. There seems to be some BR/EDR tests under tests/bluetooth/classic but I've not figured out how to run them yet.

Thalley commented:
> > I understand what you are trying to achieve here, and I'm not opposed to it as such.
> > However your argument that ISO is more timing sensitive also applies to e.g. audio over BR/EDR, right?
> > I'm wondering if, if we continue with different priorties per connection type, that we should do it in a more generic way that also accommodates the classic connection types, and any future possible types?
> 
> Unfortunately, I only have been using Zephyr with Nordic bluetooth/ble chips and don't have any BR/EDR chip to test with and am generally not that familiar with BR/EDR, so I'm not sure what a more generic change would look like. Any suggestions?
> 
> For another PR, I added some tests under tests/bsim, but those appear to only cover BLE. There seems to be some BR/EDR tests under tests/bluetooth/classic but I've not figured out how to run them yet.

We do not really test Classic/BR/EDR in the same way as LE, and like you I don't have access to any hardware that supports it either. 

W.r.t. a generic solution, I was considering if there should be a way to denote a priority per connection type, instead of a hardcoded sequence of calls. It may be overkill if there isn't an issue with classic (@lylezhu2012 please provide your input here), but it should be possible to design such a solution. Could use something like https://github.com/zephyrproject-rtos/zephyr/pull/79855 as inspiration

alwa-nordic commented:
I'm for this change. It was the intention all along[^1] that we can tweak the connection TX scheduler.

But I'm against introducing the Kconfig. IMO, just make this the new behavior without any config. I think we can safely say that ISO data has higher priority than ACL. At least it's good approximation for now. RFC.

I think we can make a good one-size-fits-all scheduler. We do not guarantee that the scheduler does not change, so we can continue to tweak it. Adding a Kconfig will make tweaking it more difficult. Default `y` Kconfigs are especially cumbersome to change[^2]. Every Kconfig is a new feature, and I think we need to see the need for a feature before we add it.

@Thalley is probably right that we should have an even more advanced scheduler. But I think we can accept this change and still welcome a completely new implementation down the line. The second list added in this PR is a very localized change that will be easy exchange for a better solution at any time.

If a very small minority of users needs to modify the behavior of the scheduler, then I would prefer to let those users hook the scheduler and completely specify its behavior in code. If a lot of users have conflicting needs, then we will look into it when we know more.

[^1]: Since we went from each l2cap channel having its own work item without any coordination, to an explicit connection TX scheduler that has a place to have logic to pick the next work.

[^2]: So if we want a Kconfig, make a different non-default one, like "Treat ACL and ISO channels with equal priority".

mjchen0 commented on subsys/bluetooth/host/conn.c:
```diff
@@ -972,6 +976,26 @@ struct bt_conn *get_conn_ready(void)
+#if defined(CONFIG_BT_ISO)
+	struct bt_conn *conn = get_conn_ready_from_sys_slist(&bt_dev.le.iso_conn_ready);
+
+	if (conn) {
```
done

Thalley commented on subsys/bluetooth/host/conn.c:
```diff
@@ -972,6 +976,27 @@ struct bt_conn *get_conn_ready(void)
+	}
+
+#if defined(CONFIG_BT_ISO)
+	/* check for iso connections first because iso is time sensitive */
```
```suggestion
	/* Check for ISO connectionss first because ISO can be time sensitive */
```

Thalley commented on subsys/bluetooth/host/conn.c:
```diff
@@ -866,8 +866,20 @@ void bt_conn_data_ready(struct bt_conn *conn)
+			sys_slist_append(&bt_dev.le.conn_ready,
+					 &conn->_conn_ready);
+		}
+#else
```
```suggestion
#else /* !CONFIG_BT_ISO */
```

mjchen0 replied:
Done

Thalley commented on subsys/bluetooth/host/conn.c:
```diff
@@ -866,8 +866,20 @@ void bt_conn_data_ready(struct bt_conn *conn)
+#else
 		sys_slist_append(&bt_dev.le.conn_ready,
 				 &conn->_conn_ready);
+#endif
```
```suggestion
#endif /* CONFIG_BT_ISO */
```

Thalley reviewed (commented):
Some minor comments, otherwise LGTM. 

The BSIM CI issue is interesting, but pretty hard to debug just going by the logs without https://github.com/zephyrproject-rtos/zephyr/pull/90039. 

Will have to try to pull down both PRs and run the test manually to see what causes this issue.

mjchen0 commented:
> Some minor comments, otherwise LGTM.
> 
> The BSIM CI issue is interesting, but pretty hard to debug just going by the logs without #90039.
> 
> Will have to try to pull down both PRs and run the test manually to see what causes this issue.

I tried to take a look at the BSIM CI failure and I think what's going on is that if the number of TX buffers (which is also the number of TX contexts) is too small, the ACL channel might get starved and management of the ISO connection fails. That's just a guess because I didn't trace it all the way through.

Notice that at the start of the test, there are these warnings from hci_core.c:

```
d_01: @00:00:00.005862  (CPU:0): [00:00:00.005,859] <wrn> bt_hci_core: Num of Controller's ACL packets != ACL bt_conn_tx contexts (7 != 3)
1480
d_01: @00:00:00.005862  (CPU:0): [00:00:00.005,859] <wrn> bt_hci_core: Num of Controller's ISO packets != ISO bt_conn_tx contexts(8 != 4)
1481
d_00: @00:00:00.005862  (CPU:0): [00:00:00.005,859] <wrn> bt_hci_core: Num of Controller's ACL packets != ACL bt_conn_tx contexts (7 != 3)
1482
d_00: @00:00:00.005862  (CPU:0): [00:00:00.005,859] <wrn> bt_hci_core: Num of Controller's ISO packets != ISO bt_conn_tx contexts(8 != 2)
```

These are existing warnings unrelated to my change, but seems to indicate that the host side is configured with fewer buffers than the controller.

If I change CONFIG_BT_ISO_TX_BUF_COUNT to 8 and CONFIG_BT_BUF_ACL_TX_COUNT to 7 for both the client and server, the warnings go away and the test completes/passes.

However, since the test passed without my CL, I guess it is still considered a regression? If for some reason someone wanted to configure the host with far fewer buffers than the controller, or if there was so much ISO traffic that all the buffers and tx context get used up such that none is available for ACL, there could still be a problem.

Not sure how to address possible starvation. Maybe reserve a tx context for ACL and don't let ISO use them all? Seems possibly fragile still.

Thalley commented on subsys/bluetooth/host/conn.c:
```diff
@@ -972,6 +976,27 @@ struct bt_conn *get_conn_ready(void)
+	}
+
+#if defined(CONFIG_BT_ISO)
+	/* check for ISO connections first because ISO is time sensitive */
```
```suggestion
	/* check for ISO connections first because ISO can be time sensitive */
```

A minor correction - Not all ISO data carries a timestamp or is time sensitive :)

Thalley commented:
> Not sure how to address possible starvation. Maybe reserve a tx context for ACL and don't let ISO use them all? Seems possibly fragile still.

Hmm, yeah, we do implement a risk of ACL starvation now, if there is always ISO data to be sent. The ISO SDU interval can be as low as 255us, so in extreme cases ACL would never get a chance. 

A naive approach is to always check for ACL data every X TX (e.g. every 10th TX we check for ACL first), but I don't like that approach. The current approach of a simple FIFO for all data ensured that all connections got their time.

Thalley reviewed (changes requested):
We need to consider the starvation issue before merging

PavelVPV commented:
> If for some reason someone wanted to configure the host with far fewer buffers than the controller, or if there was so much ISO traffic that all the buffers and tx context get used up such that none is available for ACL, there could still be a problem.

Why would someone need this? This is hundreds of unutilized bytes.

mjchen0 commented on samples/bluetooth/bap_unicast_client/prj.conf:
```diff
@@ -8,7 +8,8 @@ CONFIG_BT_GATT_AUTO_DISCOVER_CCC=y
 CONFIG_BT_BAP_UNICAST_CLIENT=y
-CONFIG_BT_ISO_TX_BUF_COUNT=4
+CONFIG_BT_ISO_TX_BUF_COUNT=8
+CONFIG_BT_BUF_ACL_TX_COUNT=7
```
I was able to get the test to pass without changing the test prj.conf.  I'll keep fixing all samples that use nrf5340_cpunet_iso-bt_ll_sw_split.conf a separate exercise for someone else to tackle, or I could try doing so in a separate CL/PR.

Thalley commented:
> After thinking some more about the starvation issue, I made an additional change which I think should address it. The current code has separate pools of bt_conn_tx for ISO and non-ISO, but they are combined into one free list. I made an additional change so that we now keep to separate free lists of bt_conn_tx, one for ISO and one for non-ISO. So, the tx_processor will prioritize sending ready ISO data, but eventually will run out of ISO bt_conn_tx. Then it will naturally stop sending ready ISO data until outstanding ones are freed, and the tx_processor will switch to processing any non-ISO ready connections.
> 
> The test passes now without any adjustment to buffer counts.

The current test failure may not be related to your PR, but perhaps due to some timing. It attempts to send an SDU every 9900us with an SDU interval of 10000us:
```
k_work_schedule(&iso_send_work, K_USEC(9900));
```

Could be that your PR changes the timing so that after a few hundred SDUs it's missing one. 
It could also be that it's hitting this:
```c
	buf = net_buf_alloc(&tx_pool, K_MSEC(BUF_ALLOC_TIMEOUT_MS));
	if (!buf) {
		/* Blocking in system work qeueue causes deadlock, hence yield
		 * and try getting the buffer again.
		 */
		if (retry_yield_wq) {
			retry_yield_wq--;

			k_work_schedule(&iso_send_work, K_USEC(BUF_ALLOC_RETRY_TIMEOUT_US));
		}
```

Which unfortunately isn't visible from the logs :(

Thalley commented:
I'm starting to wonder about the usefullness of the suggest change as well. 
 
The commit claims that "ISO is time sensitive", which is definitely the case for e.g. Audio, but may not be if you just have a sensor that outputs the tire pressure every minute
The claim also implies that ACL (or other type of connections) are not time sensitive. As I mentioned in this PR, A2DP in classic would likely be as time sensitive as LE Audio over ISO, and there might be other profiles/services/usecases such as HID that is equally, if not more, time sensitive than ISO. 
 
I think the claim about time sensitivity is not always true, and it may depend entirely on the usecase. Perhaps the current round-robin approach is actually better, unless we want to start implementing some advanced scheduling in the host, but that would probably require significant effort. Basically, given that you can have a audio over one ISO channel that is arguably time-sensitive, and a sensor sending data over another non-time sensitive channel, then maybe each connection object or each SDU should have a priority.

mjchen0 committed:
tests: bsim: bluetooth: move iso send to dedicated workq thread

With much of the bluetooth stack processing now done on the
system workq, having the test app do iso sending on the system
the same workq thread can run into problems.

Since the iso work is scheduled slightly earlier than the
ISO connection interval (9.9ms vs 10ms), and there is
only one ISO net_buf, eventually the net_buf_alloc() will
block. If the iso sending blocks for 10ms on the system workq,
the handling of the HCI response for sending the previous
iso buffer is actually blocked as well, preventing the buffer
from being released when it could have been. This then causes
causing no data to be available for the next iso event, so
then BIS sendis a zero length packet just to keep timing,
and the receiver eventually errors out when it receives the
zero length packet because the sequence number in the packet
wasn't the expected value.

The problem can be avoided in a number of ways, like changing
timeouts and retries, but I went with moving the iso work
to its own workq thread. The way the test tries to mitigate
for clock differences between the bluetooth controller and
the host seems a bit complicated and I didn't want to rewrite it.

Signed-off-by: Mike J. Chen <mjchen@google.com>

Thalley commented:
> That's why my original version had a config option to enable/disable the new logic, but it was suggested to make it always enabled. Should I bring back the config option so it can be enabled or disabled by the app? Would the preference be to keep it disabled by default?

I suggest you wait for any updates to this PR until we can agree on which direction to go. 

@rugeGerritsen you gave my comment a thumbs-up - Do you have anything additional to add? 

Would also like to get input from @alwa-nordic when they have time to review this again :) @cvinayak may also have some input
